### PR TITLE
Add recursive pattern to handle nested objects properly.

### PR DESCRIPTION
**Pull request by bardzusny**
Basic test case (little too crafty? - comments welcome) and actual implementation.

Related issue: #19 .

(note about template strings use: I was tempted to re-write whole mixin to use ES6, finally limited it to stay within the scope of what I'm actually doing - will be happy to extend this approach either way, i.e. either go back to ES5 for everything or actually go ahead and convert everything to newer ECMAScript)

### CODE REVIEW

**Comment by bardzusny** [comment]
Failing Travis build - due to general problem with addon setup :( . If this will block the merge - I'm happy to do another PR fixing this (but likely updating whole stack/addon structure to use latest `ember-cli` at the same time).

**Comment by funtusov** [comment]
Hi Adrian, thanks a lot for your work that's very helpful! I was planning to fix the setup before merging new PR's (it's from quite ancient ember-cli times indeed), if you could do that – it would be extremely helpful and we'll be able to merge this quickly

**Comment by bardzusny** [comment]
Hello @funtusov - did just that in #21. Once it's merged I'll sync this branch with master, then I think we'll be good to go. 👍

**Comment by GCorbel** [comment]
It works for me. Thanks @funtusov and @bardzusny .

File: addon/mixins/form-data-adapter.js [hunk 1]
```diff
@@ -28,20 +28,25 @@ export default Ember.Mixin.create({
...
     return formData;
-  }
+  },
+
+  _appendValue(value, formKey, formData) {
+    if (Ember.isArray(value)) {
+      value.forEach(function(item) {
+        this._appendValue(item, `${formKey}[]`, formData);
+      }, this);
+    } else if (typeof value === 'object' && value !== null) {
```
Comment (GCorbel) [inline addon/mixins/form-data-adapter.js]: I had to add ` && value.constructor.name !== 'File'` to this line.

Comment (bardzusny) [inline addon/mixins/form-data-adapter.js]: @GCorbel that's actually correct - and seems like a blunder on my side. I used leaned-down version of modifications proposed in this PR in the project I'm currently working on, apparently this check is something I missed when moving changes over. Will check this twice and add test case to be sure - I'm more and more leaning towards handling just plain objects here (`Object` instances).

**Comment by GCorbel** [comment]
There is a problem with ember-cli-mirage. The addon is waiting for a json and is not able to do a `JSON.parse` with a `FormData`.

**Comment by bardzusny** [comment]
@GCorbel regarding your earlier comment (https://github.com/funtusov/ember-cli-form-data/pull/20#discussion_r89852945) - I added additional check to make sure we're iterating over keys of only plain objects (not files). Pretty sure everything's in place now.

Regarding `ember-cli-mirage` not handling FormData properly - don't think there's much that we can do from the side of this addon . Check out this issue: https://github.com/samselikoff/ember-cli-mirage/issues/74 .

**Comment by funtusov** [comment]
@bardzusny Hi Adrian! I've merged #21, thanks a lot for the updates. Could you please rebase this PR now?

**Comment by bardzusny** [comment]
@funtusov just did that - though my latest commit broke TravisCI build. Apparently PhantomJS doesn't support `File` constructor, which I'm using in tests to confirm that only POJOs properties are iterated over. I'll take a look and fix that soon (with switch to Chrome probably on the cards).

**Comment by funtusov** [comment]
Amazing work, Adrian. One last thing: could you please add an example of usage with nested objects to the readme file?

**Comment by bardzusny** [comment]
@funtusov thanks!

About describing this feature in the README: I'm not exactly sure what's the best to go around with that, and whether it's even necessary, actually. To demonstrate it we'd have to:

* Define 2 separate Ember Data models, having `belongsTo()` relationship to one another (let's continue with `User` and `UserData` as an example, `UserData` having `avatar` file field).
* Define custom adapter for `User` model, using `FormDataAdapterMixin` from this addon.
* Define custom serializer for `User` model, using `EmbeddedRecordsMixin` and defining relationship to `userData` as embedded attribute.

... which simply seems like too much material to force into documentation of this addon. I rather expected this feature (recursive object serialization) to be something that works out-of-the-box, and seems very natural - so perhaps not mentioning it explicitly is not a bad idea, and at most short note (that "_fields on embedded relationships are serialized properly_") would suffice? WDYT?

**Comment by funtusov** [comment]
Ok, I agree with your point, merging

**Comment by funtusov** [comment]
Published `ember-cli-form-data@1.0.0`

**Pull request merged**